Add Paging 3.0 to Sunflower

tiembo opened the pull request.

tiembo committed:
Add Paging dependency

Also updates RecyclerView version and update ShareCompat API call

tiembo committed:
Update UnsplashService param order

Move client_id to the last param

tiembo committed:
Implement Paging 3.0

Also remove default values for UnsplashService#searchPhotos

tiembo committed:
Remove unused method

manuelvicnt commented on app/src/main/java/com/google/samples/apps/sunflower/viewmodels/GalleryViewModel.kt:
```diff
@@ -16,30 +16,29 @@
-            val repos = repository.getSearchResultStream(queryString).asLiveData()
-            emitSource(repos)
+    fun searchPictures(queryString: String): Flow<PagingData<UnsplashPhoto>> {
+        val lastResult = currentSearchResult
```
`lastResult` seems like can be avoidable? Maybe just use `currentSearchResult` that IMO is easier to understand in the logic below

tiembo replied:
Good catch!

manuelvicnt commented on app/src/main/java/com/google/samples/apps/sunflower/data/UnsplashPagingSource.kt:
```diff
@@ -0,0 +1,47 @@
+                prevKey = if (page == UNSPLASH_STARTING_PAGE_INDEX) null else page - 1,
+                nextKey = if (page == response.totalPages) null else page + 1
+            )
+        } catch (exception: IOException) {
```
Maybe catch `exception: Exception`? Otherwise it might throw if something else happens

tiembo replied:
Yeah, that's safer

manuelvicnt reviewed (commented):
Florina to review the Paging part but apart from that, this LGTM :) thanks

florina-muntenescu commented on app/src/main/java/com/google/samples/apps/sunflower/adapters/GalleryAdapter.kt:
```diff
@@ -41,7 +41,9 @@ class GalleryAdapter : ListAdapter<UnsplashPhoto, GalleryViewHolder>(GalleryDiff
     override fun onBindViewHolder(holder: GalleryViewHolder, position: Int) {
         val photo = getItem(position)
-        holder.bind(photo)
+        if (photo != null) {
```
Paging will return null items as placeholders
We should either support them or in `PagingConfig`, disable the placeholders

tiembo replied:
Disabled the placeholders

florina-muntenescu commented on build.gradle:
```diff
@@ -40,7 +40,8 @@ buildscript {
         navigationVersion = '2.2.0'
         okhttpLoggingVersion = '4.7.2'
-        recyclerViewVersion = '1.1.0-alpha05'
+        pagingVersion = '3.0.0-alpha01'
```
```suggestion
        pagingVersion = '3.0.0-alpha02'
```

florina-muntenescu commented on build.gradle:
```diff
@@ -40,7 +40,8 @@ buildscript {
         okhttpLoggingVersion = '4.7.2'
-        recyclerViewVersion = '1.1.0-alpha05'
+        pagingVersion = '3.0.0-alpha01'
+        recyclerViewVersion = '1.2.0-alpha03'
```
```suggestion
        recyclerViewVersion = '1.2.0-alpha04'
```

tiembo committed:
Update paging version

Co-authored-by: Florina Muntenescu <2998890+florina-muntenescu@users.noreply.github.com>

tiembo committed:
Update recyclerview version

Co-authored-by: Florina Muntenescu <2998890+florina-muntenescu@users.noreply.github.com>

tiembo committed:
Address PR comments

- catch Exception instead of IO/HttpException
- disable Paging placeholders
- remove unneeded local variable in searchPictures

tiembo commented:
Closing discussion PR - merged into upstream
